sympy/integrals/laplace.py

```diff
@@ -593,46 +583,243 @@ def _laplace_rule_delta(f, t, s):
+                nc[k] = nc[k].rewrite(cos)
+            else:
+                nc[k] = (ri[0] + I*ri[1]).rewrite(cos)
+        return nc
+
+    def _quadpole(t1, k1, k2, k3, s):
+        a, k0, a_r, a_i = t1['a'], t1['k'], t1[re], t1[im]
+        nc = [
+            k0 + k1 + k2 + k3,
+            a*(k0 + k1 - k2 - k3) - 2*I*a_i*k1 + 2*I*a_i*k2,
+            (
+                a**2*(-k0 - k1 - k2 - k3) +
+                a*(4*I*a_i*k0 + 4*I*a_i*k3) +
+                4*a_i**2*k0 + 4*a_i**2*k3),
+            (
+                a**3*(-k0 - k1 + k2 + k3) +
+                a**2*(4*I*a_i*k0 + 2*I*a_i*k1 - 2*I*a_i*k2 - 4*I*a_i*k3) +
+                a*(4*a_i**2*k0 - 4*a_i**2*k3))
+            ]
+        dc = [
+            S.One, S.Zero, 2*a_i**2 - 2*a_r**2,
+            S.Zero, a_i**4 + 2*a_i**2*a_r**2 + a_r**4]
+        n = Add(
+            *[x*s**y for x, y in zip(_simpc(nc), range(len(nc))[::-1])])
+        d = Add(
+            *[x*s**y for x, y in zip(dc, range(len(dc))[::-1])])
+        debugf('        quadpole: (%s) / (%s)', (n, d))
+        return n/d
+
+    def _ccpole(t1, k1, s):
+        a, k0, a_r, a_i = t1['a'], t1['k'], t1[re], t1[im]
+        nc = [k0 + k1, -a*k0 - a*k1 + 2*I*a_i*k0]
+        dc = [S.One, -2*a_r, a_i**2 + a_r**2]
+        n = Add(
+            *[x*s**y for x, y in zip(_simpc(nc), range(len(nc))[::-1])])
+        d = Add(
+            *[x*s**y for x, y in zip(dc, range(len(dc))[::-1])])
+        debugf('        ccpole: (%s) / (%s)', (n, d))
+        return n/d
+
+    def _rspole(t1, k2, s):
+        a, k0, a_r, a_i = t1['a'], t1['k'], t1[re], t1[im]
+        nc = [k0 + k2, a*k0 - a*k2 - 2*I*a_i*k0]
+        dc = [S.One, -2*I*a_i, -a_i**2 - a_r**2]
+        n = Add(
+            *[x*s**y for x, y in zip(_simpc(nc), range(len(nc))[::-1])])
+        d = Add(
+            *[x*s**y for x, y in zip(dc, range(len(dc))[::-1])])
+        debugf('        rspole: (%s) / (%s)', (n, d))
+        return n/d
+
+    def _sypole(t1, k3, s):
+        a, k0 = t1['a'], t1['k']
+        nc = [k0 + k3, a*(k0 - k3)]
+        dc = [S.One, S.Zero, -a**2]
+        n = Add(
+            *[x*s**y for x, y in zip(_simpc(nc), range(len(nc))[::-1])])
+        d = Add(
+            *[x*s**y for x, y in zip(dc, range(len(dc))[::-1])])
+        debugf('        sypole: (%s) / (%s)', (n, d))
+        return n/d
+
+    def _simplepole(t1, s):
+        a, k0 = t1['a'], t1['k']
+        n = k0
+        d = s - a
+        debugf('        simplepole: (%s) / (%s)', (n, d))
+        return n/d
+
+    while len(xm) > 0:
+        t1 = xm.pop()
+        i_imagsym = None
+        i_realsym = None
+        i_pointsym = None
+        # The following code checks all remaining poles. If t1 is a pole at
+        # a+b*I, then we check for a-b*I, -a+b*I, and -a-b*I, and
+        # assign the respective indices to i_imagsym, i_realsym, i_pointsym.
+        # -a-b*I / i_pointsym only applies if both a and b are != 0.
+        for i in range(len(xm)):
```

Comment:
This double loop is `O(n**2)`.

Reply:
Worst case `N**2/2`, yes, but can it be done in less? I cannot imagine a way in which we do not have, in the worst case, to compare every pole to every pole, if we have single poles. Using sets would just hide the fact.

Reply:
No using sets does not hide the fact. Both sets and dicts have `O(1)` lookup. Here is a demo of `O(n)` using sets and dicts:
```python
from collections import defaultdict

def groupby(items, keyfunc):
    """Group items having keyfunc(item1) == keyfunc(item2)"""
    key_to_items = defaultdict(set)
    for item in items:
        key_to_items[keyfunc(item)].add(item)
    return list(key_to_items.values())

def match_neg_pairs(items, keyfunc):
    """Find pairs having keyfunc(item1) == -keyfunc(item2)"""
    key_to_items = {keyfunc(item): item for item in items}
    key_vals = {keyfunc(item) for item in items}
    neg_key_vals = {-keyfunc(item) for item in items}

    matched_keys = key_vals & neg_key_vals
    unmatched_keys = key_vals - neg_key_vals

    unmatched_items = [key_to_items[key] for key in unmatched_keys]
    
    matched_pairs = []
    for key in matched_keys:
        if key not in key_to_items:
            continue
        item1 = key_to_items.pop(key)
        item2 = key_to_items.pop(-key)
        matched_pairs.append((item1, item2))

    return matched_pairs, unmatched_items


items = {1+I, 1-I, 2+I, 3+I, -1+I, -1-I, 4+I, 4-I}

cc_pairs = []
unpaired = []

for items_re in groupby(items, re):
    cc_pairs_re, unpaired_re = match_neg_pairs(items_re, im)
    cc_pairs.extend(cc_pairs_re)
    unpaired.extend(unpaired_re)

quad_pairs, cc_pairs = match_neg_pairs(cc_pairs, lambda p: re(p[0]))

print(quad_pairs)
print(cc_pairs)
print(unpaired)
```

Reply:
That needs improvement but the basic idea is there that we use dicts and sets to avoid all-to-all comparison.

Reply:
I am confused. If `match_neg_pairs` contains `O(n)` functions, and then you loop through the items with `for items_re in groupby(items, re):`, how is that not `O(n**2)`?

Reply:
P.S. If I find time later I will profile both versions.

Reply:
I could not resist.  Yours is also `O(n**2)` and a factor 30 slower.

```
    31        **14**         96.0      6.9      0.9      for key in matched_keys:

Total time: 0.0096572 s
Function: dict_it at line 40

   68        **16**         91.0      5.7      2.8          for i in range(len(xm)):

Total time: 0.0003295 s
File: 0038_compareloops.py
Function: loop_it at line 53
```

[0038.zip](https://github.com/sympy/sympy/files/10707560/0038.zip)

Reply:
Maybe I am drawing the wrong conclusions here?

Reply:
> I am confused. If `match_neg_pairs` contains `O(n)` functions, and then you loop through the items with `for items_re in groupby(items, re):`, how is that not `O(n**2)`?

The `for items_re` loop is going over a partition of the items:
```
In [8]: items
Out[8]: {-1 + ⅈ, 1 - ⅈ, 1 + ⅈ, 2 + ⅈ, 3 + ⅈ}

In [9]: groupby(items, re)
Out[9]: [{1 - ⅈ, 1 + ⅈ}, {2 + ⅈ}, {3 + ⅈ}, {-1 + ⅈ}]
```
We're just splitting the list into several smaller lists. Then `match_neg_pairs` loops over each smaller list but that's still `O(n)` where `n` is the size of the original larger list.

You need to be careful with profiling things in SymPy because of the cache. Here the bulk of the time is spent in the `re` function but the results of calling `re` are cached for subsequent calls. That means that if you call `dict_it` before `loop_it` then it will look like `dict_it` takes longer but actually the time is just being spent in `re`. Try reversing the order in which you call the two functions.

Also to see the effects of quadratic runtime you really need to make the input larger e.g.:
```
items = {n + I for n in range(100)}
```
Probably things need to get quite large here before the quadratic effects come into play because there is a large linear cost just in computing the real parts (but ideally we would improve that by speeding things up elsewhere).

In any case that code can be improved. What I really want is for this to be more readable but looking at it now it is more readable with the comments. A large loop that modifies multiple data structures is difficult to follow so I think it's important to do what we can to shrink that loop as much as possible but maybe it isn't really possible to do that here.

Reply:
> You need to be careful with profiling things in SymPy because of the cache.

I tried it, and so it is. What an ugly trap.

> In any case that code can be improved. What I really want is for this to be more readable but looking at it now it is more readable with the comments.

Shall we leave it for now, then? I cannot say that I can really read your code.

Reply:
> Shall we leave it for now, then?

Yes, I think so.